assets/js/interactivity/store.ts

```diff
@@ -0,0 +1,240 @@
+/**
+ * External dependencies
+ */
+import { deepSignal } from 'deepsignal';
+import { computed } from '@preact/signals';
+import {
+	resetScope,
+	getScope,
+	setScope,
+	setNamespace,
+	resetNamespace,
+} from './hooks';
+
+const isObject = ( item ) =>
+	item && typeof item === 'object' && ! Array.isArray( item );
+
+const deepMerge = ( target, source ) => {
+	if ( isObject( target ) && isObject( source ) ) {
+		for ( const key in source ) {
+			if (
+				typeof Object.getOwnPropertyDescriptor( source, key )?.get ===
+				'function'
+			) {
+				Object.defineProperty( target, key, {
+					get: Object.getOwnPropertyDescriptor( source, key ).get,
+				} );
+			} else if ( isObject( source[ key ] ) ) {
+				if ( ! target[ key ] ) Object.assign( target, { [ key ]: {} } );
+				deepMerge( target[ key ], source[ key ] );
+			} else {
+				Object.assign( target, { [ key ]: source[ key ] } );
+			}
+		}
+	}
+};
+
+const parseInitialState = () => {
+	const storeTag = document.querySelector(
+		`script[type="application/json"]#wc-interactivity-initial-state`
+	);
+	if ( ! storeTag?.textContent ) return {};
+	try {
+		const initialState = JSON.parse( storeTag.textContent );
+		if ( isObject( initialState ) ) return initialState;
+		throw Error( 'Parsed state is not an object' );
+	} catch ( e ) {
+		// eslint-disable-next-line no-console
+		console.log( e );
+	}
+	return {};
+};
+
+export const stores = new Map();
+
+const objToProxy = new WeakMap();
+const proxyToNs = new WeakMap();
+
+const proxify = ( obj: any, ns: string ) => {
+	if ( ! objToProxy.has( obj ) ) {
+		const proxy = new Proxy( obj, handlers );
+		objToProxy.set( obj, proxy );
+		proxyToNs.set( proxy, ns );
+	}
+
+	return objToProxy.get( obj );
+};
+
+const handlers = {
+	get: ( target, key, receiver ) => {
+		const ns = proxyToNs.get( receiver );
+
+		// Check if the proxy is the store root and no prop with that name
+		// exist. In that case, return an empty object for the prop requested.
+		if ( receiver === stores.get( ns ) && ! ( key in target ) ) {
+			const obj = {};
+			target[ key ] = obj;
+			return proxify( obj, ns );
+		}
+
+		// Check if the property is a getter.
+		const getter = Object.getOwnPropertyDescriptor( target, key )?.get;
+		if ( getter ) {
+			const scope = getScope();
+			if ( scope ) {
+				scope.getters = scope.getters || new Map();
+				if ( ! scope.getters.has( getter ) ) {
+					scope.getters.set(
+						getter,
+						computed( () => {
+							setNamespace( ns );
+							setScope( scope );
+							const result = getter.call( target );
+							// resetScope(); // maybe scope should be a stack?
+							resetNamespace();
+							return result;
+						} )
+					);
+				}
+				return scope.getters.get( getter ).value;
```

Comment:
Note: We are going to need a way to access the computed when we optimize `wp-text` and `wp-bind` to pass signals down. It doesn't have to be a nice API, so maybe something as simple as `getComputed(obj, "prop);`